flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStore.java

```diff
@@ -0,0 +1,310 @@
+		}
+
+		LOG.debug("Added {} to {}.", checkpoint, path);
+		assert (checkpointStateHandles.getLast().equals(stateHandle));
+	}
+
+	@Override
+	public CompletedCheckpoint getLatestCheckpoint() throws Exception {
+		if (checkpointStateHandles.isEmpty()) {
+			return null;
+		}
+		else {
+			return checkpointStateHandles.getLast().getState(userClassLoader);
+		}
+	}
+
+	@Override
+	public List<CompletedCheckpoint> getAllCheckpoints() throws Exception {
+		List<CompletedCheckpoint> checkpoints = new ArrayList<>(checkpointStateHandles.size());
+
+		for (ZooKeeperStateHandle<CompletedCheckpoint> stateHandle
+				: checkpointStateHandles) {
+			checkpoints.add(stateHandle.getState(userClassLoader));
+		}
+
+		return checkpoints;
+	}
+
+	@Override
+	public int getNumberOfRetainedCheckpoints() {
+		return checkpointStateHandles.size();
+	}
+
+	@Override
+	public void discardAllCheckpoints() throws Exception {
+		for (ZooKeeperStateHandle<CompletedCheckpoint> checkpoint
+				: checkpointStateHandles) {
+			try {
+				removeFromZooKeeperAndDiscardCheckpoint(checkpoint);
+			}
+			catch (Exception e) {
+				LOG.error("Failed to discard checkpoint.", e);
+			}
+		}
+
+		checkpointStateHandles.clear();
+
+		String path = "/" + client.getNamespace();
+
+		LOG.info("Removing {} from ZooKeeper", path);
+		ZKPaths.deleteChildren(client.getZookeeperClient().getZooKeeper(), path, true);
+	}
+
+	/**
+	 * Removes the state handle from ZooKeeper, discards the checkpoints, and the state handle.
+	 */
+	private void removeFromZooKeeperAndDiscardCheckpoint(
+			final ZooKeeperStateHandle<CompletedCheckpoint> stateHandle) throws Exception {
+
+		final BackgroundCallback callback = new BackgroundCallback() {
+			@Override
+			public void processResult(CuratorFramework client, CuratorEvent event) throws Exception {
+				try {
+					if (event.getType() == CuratorEventType.DELETE) {
+						if (event.getResultCode() == 0) {
+							// The checkpoint
+							CompletedCheckpoint checkpoint = stateHandle
+									.getState(userClassLoader);
+
+							checkpoint.discard(userClassLoader);
+
+							// Discard the state handle
+							stateHandle.discardState();
+
+							// Discard the checkpoint
+							LOG.debug("Discarded " + checkpoint);
+						}
+						else {
+							throw new IllegalStateException("Unexpected result code " +
+									event.getResultCode() + " in '" + event + "' callback.");
+						}
+					}
+					else {
+						throw new IllegalStateException("Unexpected event type " +
+								event.getType() + " in '" + event + "' callback.");
+					}
+				}
+				catch (Exception e) {
+					LOG.error("Failed to discard checkpoint.", e);
+				}
+			}
+		};
+
+		// Remove state handle from ZooKeeper first. If this fails, we can still recover, but if
+		// we remove a state handle and fail to remove it from ZooKeeper, we end up in an
+		// inconsistent state.
+		checkpointsInZooKeeper.remove(stateHandle.getPathInZooKeeper(), callback);
```

Comment:
I think this requires to much knowledge of the way you want to remove the state from ZooKeeper. I think the ZooKeeperStateHandle was not well thought it. I will remove it and manage (StateHandle, String) pairs in CheckpointStore manually.